### PR TITLE
Reland: Add OrderedFocusTraversalPolicy and FocusTraversalGroup (#49235) 

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This re-lands #49235 with the addition of `includeSemantics` flag on the `Focus` widget so that the `FocusTraversalGroup` can create a `Focus` widget without affecting the semantics tree.

The `FocusTraversalGroup` uses the `Focus` widget to create a grouping of descendants for traversal, but doesn't actually participate in focus (`canRequestFocus` is always false), so we don't want it to add a `Semantics` widget in that case, since that can cause semantics changes.  The `canRequestFocus` attribute can also be used when a widget is disabled, so we do sometimes want to include `Semantics` even if that is false, but not in the case where it is always false, as for `FocusTraversalGroup`.

## Tests

- Added a test to make sure that `FocusTraversalGroup` doesn't add any semantics information.

## Breaking Change

- [X] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/focus_scope.dart [hunk 1]
```diff
@@ -459,15 +472,19 @@ class _FocusState extends State<Focus> {
...
+    Widget semantics;
+    if (widget.includeSemantics) {
+      semantics = Semantics(
+        focusable: _canRequestFocus,
+        focused: _hasPrimaryFocus,
+        child: widget.child,
+      );
+    } else {
+      semantics = widget.child;
+    }
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/focus_scope.dart]: In many other places of the framework we use this pattern for these kinda things:

```dart
Widget child = widget.child;
if (widget.includeSemantics) {
  child = Semantics(
     // ...
    child: child,
   );
}
```

**Review by goderbauer — COMMENTED** [review]
LGTM

**Comment by goderbauer** [comment]
Looks like cirrus is unhappy, though.

File: packages/flutter/lib/src/widgets/focus_scope.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/focus_scope.dart]: Thanks.  Updated to match that pattern.

**Comment by hterkelsen** [comment]
Renaming `WidgetOrderFocusTraversalPolicy` to `WidgetOrderTraversalPolicy` is a breaking change. It causes [the gallery app](https://github.com/flutter/samples/tree/master/gallery) to fail to build because it uses `WidgetOrderFocusTraversalPolicy`.

Here is a sample error message I see when trying to build the gallery with the latest master:

```
Target dart2js failed: Exception: lib/pages/home.dart:1121:48:
Error: Type 'WidgetOrderFocusTraversalPolicy' not found.
class StudyWrapperFocusTraversalPolicy extends WidgetOrderFocusTraversalPolicy {
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/layout/focus_traversal_policy.dart:13:48:
Error: Type 'WidgetOrderFocusTraversalPolicy' not found.
class EdgeChildrenFocusTraversalPolicy extends WidgetOrderFocusTraversalPolicy {
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by gspencergoog** [comment]
@hterkelsen, does the gallery app have a test in [flutter/tests](https://github.com/flutter/tests)?

If not, it probably should.  According to our [Breaking Change](https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes) policy, unless it breaks a test in that repo or one of our tests, it's not considered a breaking change.

**Comment by johnsonmh** [comment]
@gspencergoog, since you had to update `packages/flutter/test/widgets/focus_scope_test.dart`, doesn't this count as a breaking change?

**Comment by gspencergoog** [comment]
No, because I didn't have to make that change to keep code from breaking.  I only had to make that change because I wanted to avoid the warnings from deprecating `DefaultFocusTraversal`, and our policy explicitly said that deprecations weren't breaking changes.  The wording of the policy has since changed a bit.

**Comment by johnsonmh** [comment]
Good to know, thanks for clarifying!

**Pull request merged**